### PR TITLE
Fixed #23932 -- Added docs on how to migrate unique fields 

### PR DESCRIPTION
**Pull request by akulakov**

### CODE REVIEW

**Comment by timgraham** [comment]
Could you try to fix your branch so it doesn't contain all the unrelated commits? Also, fixes -> fixed in the commit message.

File: docs/ref/migration-operations.txt [hunk 1]
```diff
@@ -365,6 +365,84 @@ It accepts two list of operations, and when asked to apply state will use the
...
+
+Applying a migration that adds unique fields to existing rows will raise an error
+because the unique field will be generated with the same value added to all rows,
+breaking the `unique` constraint.
+
+Therefore, the following steps should be taken (in this example, we'll
+create a uuid field, so modify with appropriate `default` arg for your
+needs):
+
+* add the field with `default=uuid.uuid4` and `unique=True` args.
```
Comment (timgraham) [inline docs/ref/migration-operations.txt]: Could you be consistent about whether or not each point starts with a capital letter and ends with a period (I suggest "yes" for both)?

File: docs/ref/migration-operations.txt [hunk 2]
```diff
@@ -365,6 +365,84 @@ It accepts two list of operations, and when asked to apply state will use the
...
+
+* run the `makemigrations` command
```
Comment (timgraham) [inline docs/ref/migration-operations.txt]: Wherever you have single backticks, they need to be doubled for proper rendering.

**Comment by akulakov** [comment]
Fixed things up according to Tim's comments - thanks!

File: docs/ref/migration-operations.txt [hunk 3]
```diff
@@ -365,6 +365,84 @@ It accepts two list of operations, and when asked to apply state will use the
 state list, and when asked to apply changes to the database will use the database
 list. Do not use this operation unless you're very sure you know what you're doing.
 
+Creating migrations that add unique fields
+===========================================
```
Comment (MarkusH) [inline docs/ref/migration-operations.txt]: Underline is too long

File: docs/ref/migration-operations.txt [hunk 4]
```diff
@@ -365,6 +365,84 @@ It accepts two list of operations, and when asked to apply state will use the
...
+
+Applying a migration that adds unique fields to existing rows will raise an error
```
Comment (MarkusH) [inline docs/ref/migration-operations.txt]: "... to existing non-nullable rows ..."

File: docs/ref/migration-operations.txt [hunk 5]
```diff
@@ -365,6 +365,84 @@ It accepts two list of operations, and when asked to apply state will use the
...
+because the unique field will be generated with the same value added to all rows,
```
Comment (MarkusH) [inline docs/ref/migration-operations.txt]: "... with the same default value added ..."

File: docs/ref/migration-operations.txt [hunk 6]
```diff
@@ -365,6 +365,84 @@ It accepts two list of operations, and when asked to apply state will use the
...
+breaking the ``unique`` constraint.
```
Comment (MarkusH) [inline docs/ref/migration-operations.txt]: "... constraint, if a default value is given."

File: docs/ref/migration-operations.txt [hunk 7]
```diff
@@ -365,6 +365,84 @@ It accepts two list of operations, and when asked to apply state will use the
...
+            migrations.AddField(
+                model_name='mymodel',
+                name='uuid',
+                field=models.UUIDField(max_length=32, unique=True, default=uuid.uuid4),
+            ),
+        ]
+
+* You will need to make three changes:
+
+    - Change the ``AddField`` operation to create an intermediary null
```
Comment (MarkusH) [inline docs/ref/migration-operations.txt]: Please use asterisks as bullet points.

File: docs/ref/migration-operations.txt [hunk 8]
```diff
@@ -365,6 +365,84 @@ It accepts two list of operations, and when asked to apply state will use the
...
+      field.
+
+    - Add a RunPython operation to generate a UUID for each existing row.
```
Comment (MarkusH) [inline docs/ref/migration-operations.txt]: "Add a \`\`RunPython\`\` or \`\`RunSQL\`\` operation ..."

File: docs/ref/migration-operations.txt [hunk 9]
```diff
@@ -365,6 +365,84 @@ It accepts two list of operations, and when asked to apply state will use the
...
+
+    - Add a second ``AmendField`` operation to change the field into
```
Comment (MarkusH) [inline docs/ref/migration-operations.txt]: That one is called "AlterField"

File: docs/ref/migration-operations.txt [hunk 10]
```diff
@@ -365,6 +365,84 @@ It accepts two list of operations, and when asked to apply state will use the
...
 
+Creating migrations that add unique fields
+===========================================
+
+Applying a migration that adds unique fields to existing rows will raise an error
+because the unique field will be generated with the same value added to all rows,
+breaking the ``unique`` constraint.
+
+Therefore, the following steps should be taken (in this example, we'll
+create a uuid field, so modify with appropriate ``default`` arg for your
```
Comment (MarkusH) [inline docs/ref/migration-operations.txt]: "... we'll add a non-nullable UUID field with a default value. Modify the respective field according to your needs)."

File: docs/ref/migration-operations.txt [hunk 11]
```diff
@@ -365,6 +365,84 @@ It accepts two list of operations, and when asked to apply state will use the
...
+      field.
+
+    - Add a RunPython operation to generate a UUID for each existing row.
+
+    - Add a second ``AmendField`` operation to change the field into
+      ``unique=True`` and ``null=False`` one.
+
+* The resulting migration should look similar to this::
+
+    import uuid
```
Comment (MarkusH) [inline docs/ref/migration-operations.txt]: Maybe add add comment for the default imports

File: docs/ref/migration-operations.txt [hunk 12]
```diff
@@ -365,6 +365,84 @@ It accepts two list of operations, and when asked to apply state will use the
...
+
+    def gen_uuid(apps, schema_editor):
+        from myapp.models import Mymodel
```
Comment (MarkusH) [inline docs/ref/migration-operations.txt]: Use ``MyModel = apps.get_model("myapp", "MyModel")``

File: docs/ref/migration-operations.txt [hunk 13]
```diff
@@ -365,6 +365,84 @@ It accepts two list of operations, and when asked to apply state will use the
...
+        for row in Mymodel.objects.all():
```
Comment (MarkusH) [inline docs/ref/migration-operations.txt]: ``MyModel``

File: docs/ref/migration-operations.txt [hunk 14]
```diff
@@ -365,6 +365,84 @@ It accepts two list of operations, and when asked to apply state will use the
...
+        ]
+
+        operations = [
+            migrations.AddField(
+                model_name='mymodel',
+                name='uuid',
+                field=models.UUIDField(max_length=32, null=True),
+            ),
+
+            migrations.RunPython(gen_uuid),
```
Comment (MarkusH) [inline docs/ref/migration-operations.txt]: You could add a comment referring to ``RunPython.noop`` for a ``reverse_code`` to make the migration file reversible.

File: docs/ref/migration-operations.txt [hunk 15]
```diff
@@ -365,6 +365,84 @@ It accepts two list of operations, and when asked to apply state will use the
...
+
+            migrations.AlterField(
+                model_name='mymodel',
+                name='uuid',
+                field=models.UUIDField(null=False, unique=True, default=uuid.uuid4),
+            ),
+
+        ]
+
+* At this point you can apply the migration as usual with the ``migrate`` command.
```
Comment (MarkusH) [inline docs/ref/migration-operations.txt]: I'd link to the migrate command docs. `:djadmin:\`migrate\`` I think.

**Comment by MarkusH** [comment]
You might also want to consider to link to the migration operations when you mention them for the first time.

**Comment by MarkusH** [comment]
Just saw that this change is part of the migration operations docs. I think it should be part of `docs/topics/migrations.txt` (which is getting really large)

@timgraham How do you feel about moving `docs/topics/migrations.txt` to `docs/topics/migrations/index.txt` and add `docs/topics/migrations/advanced-usecases.txt` that would for now cover this topic and maybe https://docs.djangoproject.com/en/dev/topics/migrations/#custom-fields and https://docs.djangoproject.com/en/dev/topics/migrations/#model-managers (and maybe something else as well)?

**Comment by timgraham** [comment]
Or maybe the more "recipe" type migration examples like this and "Data migrations and multiple databases" would be better suited in "howto".

**Comment by MarkusH** [comment]
:+1: @timgraham Didn't think of the howto section. That's probably the best place for that.

**Comment by akulakov** [comment]
I thought migration ops page might be better than migrations page for this because migrations page says "If you’re interested in the more advanced migration operations ..." and then links to migration ops; this recipe is arguably more advanced than things in migrations page; but mainly it's more of a special case and so less of an interest to someone reading migrations page for general overview of migrations.

Having it in Howto may be good, but it should be at least linked and shortly explained in migration ops page because of the exception that users will run into when they try this.

**Comment by MarkusH** [comment]
@akulakov Yes, the howto page should be linked from both `topics/migrations` and `ref/migration-operations`.

**Comment by akulakov** [comment]
Ok, sounds good!

**Comment by timgraham** [comment]
I don't think we should have 1 page for every example like this. As suggested above, could you first move the "Data migrations and multiple databases" example to a separate document along with an entry in the table contents and links from the other migration docs. The new content can be a separate commit after that. Not sure on a good name for the document, maybe "migration-recipes" or "writing-migrations"?

**Comment by akulakov** [comment]
@timgraham sorry but the multiple databases thing is a separate PR: https://github.com/django/django/pull/4013 ?  Or you are suggesting it should be merged into this PR? (I'm new so not sure how to best separate PRs so would be happy to get any guidance..)

**Comment by timgraham** [comment]
Sorry, this is what I was trying to describe: https://github.com/django/django/pull/4042. The idea is to merge that first and then add this in that new document.

**Comment by timgraham** [comment]
I merged the linked PR. Could you move the example to `docs/howto/writing-migrations.txt`?

**Comment by akulakov** [comment]
@timgraham Yes, thanks -- will do that.

**Comment by akulakov** [comment]
Moved it to `writing-migrations.txt`.

I think it might be good to have a direct link to this section from migrations.txt, for the reason of this use case: a user runs into the failed migration while doing a reasonable migration operation with a unique field, (and exception will refer to unique field as the source of failure), user goes to migrations.txt to look for notes on unique field migrations. If he doesn't find a specific note on that, that creates impression that either this is a bug or django docs are incomplete (because it's not obvious that this type of migration will be covered in a howto section).

Does that make sense? I was thinking of adding this like to 'see also' section:

    :doc:`Creating migrations that add unique fields to existing non-nullable rows </howto/<link>`

**Comment by timgraham** [comment]
I don't think adding every entry of the "how to" as a "see also" entry makes much sense. I think most people will use a search engine when they have a problem and will be able to find the appropriate docs.

**Comment by akulakov** [comment]
I didn't mean this should be done for every entry but only for ones (and realistically it will probably be one or two for any 'main' page), when django seems to behave in a way that could be seen as buggy (even to someone experienced with django).

In other words, when I have to do e.g. migration with multiple databases, the expectation is that there'd be a blog post or a howto section, official or otherwise, on how to do this. It seems like a "howto type of thing". With unique field migration, it seems like a regular type of migration I'd expect django to "just do".

However it's probably a fairly rare use case, so it might be ok to have people google for it.

Just thought I'd bring it up, this will also help me understand how django docs usually approach this type of "principle of least surprise" issue. thanks!

File: docs/howto/writing-migrations.txt [hunk 16]
```diff
@@ -67,3 +67,87 @@ Then, to leverage this in your migrations, do the following::
...
+
+        operations = [
+            migrations.AddField(
+                model_name='mymodel',
+                name='uuid',
+                field=models.UUIDField(max_length=32, null=True),
+            ),
+
+
+            # consider using reverse_code=RunPython.noop(apps, schema_editor) to make
```
Comment (MarkusH) [inline docs/howto/writing-migrations.txt]: * "Consider" with a capital "c".
* `reverse_code=RunPython.noop`

File: docs/howto/writing-migrations.txt [hunk 17]
```diff
@@ -67,3 +67,87 @@ Then, to leverage this in your migrations, do the following::
...
+            # migration reversible (see ``RunPython.noop()``)
```
Comment (MarkusH) [inline docs/howto/writing-migrations.txt]: I'd drop the parentheses.

File: docs/howto/writing-migrations.txt [hunk 18]
```diff
@@ -67,3 +67,87 @@ Then, to leverage this in your migrations, do the following::
...
+value added to all rows, breaking the ``unique`` constraint, if a default value is
+given.
+
+Therefore, the following steps should be taken (in this example, we'll add a
+non-nullable UUID field with a default value. Modify the respective field according
+to your needs):
+
+* Add the field with ``default=uuid.uuid4`` and ``unique=True`` args.
+
+* Run the ``makemigrations`` command.
```
Comment (MarkusH) [inline docs/howto/writing-migrations.txt]: Can you link to the command, please.

**Comment by timgraham** [comment]
This looks good. Here are some cosmetic edits: http://dpaste.com/160W9M3
One thing I thought of is that in the event someone doesn't take the database offline during the migration and a row is created after the `AddField` and before `RunPython`, the existing value for the new field will be clobbered by the migration. Do you think this warrants a note? We could try to make the migration "smarter" by trying to have the RunPython only modify rows that have the duplicate value.

**Comment by MarkusH** [comment]
@timgraham: 2 comments on your diff: line 66 has "valuae" and it looks like the uuid import got lost in the migration file.

**Comment by MarkusH** [comment]
Adding a short wanting sounds like a good idea. However, making the migration smarter would involve database level defaults from my point of view.

**Comment by akulakov** [comment]
A couple of notes (in addition to Markus' notes):

null=True is not being reset in the diff, is this intentional? I think we're setting null=True as a temporary measure to enable the operation, and if user needs null=True, he'll know not to reset it.

MyModel or Poll for model name example?

**Comment by timgraham** [comment]
Thanks for the reviews. Hopefully I fixed all the issues:  http://dpaste.com/3X53YZ3

`null=False` isn't needed in the `AlterField` because that's the default.

**Comment by akulakov** [comment]
One quick note: UUIDField is already max_length=32 by default, perhaps cleaner to leave this arg out?

**Comment by timgraham** [comment]
Yes, it's currently part of the auto-generated migration, but shouldn't be as `UUIDField` is missing a `deconstruct()` method, I'll fix that first.

**Comment by timgraham** [comment]
merged in 1f9e44030e9c5300b97ef7b029f482c53a66f13b, thanks!

Fix for `UUID.deconstuct()`: https://github.com/django/django/pull/4065

**Comment by akulakov** [comment]
Great -- much thanks for your help, @timgraham and @MarkusH !

**Pull request closed without merging**